src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectLockFileWatcher.cs

```diff
@@ -0,0 +1,165 @@
+        public ProjectLockFileWatcher([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+                                      [Import(ContractNames.ProjectTreeProviders.FileSystemDirectoryTree)] IProjectTreeProvider fileSystemTreeProvider,
+                                      IUnconfiguredProjectCommonServices projectServices,
+                                      IProjectLockService projectLockService)
+        {
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+            Requires.NotNull(fileSystemTreeProvider, nameof(fileSystemTreeProvider));
+            Requires.NotNull(projectServices, nameof(projectServices));
+            Requires.NotNull(projectLockService, nameof(projectLockService));
+
+            _serviceProvider = serviceProvider;
+            _fileSystemTreeProvider = fileSystemTreeProvider;
+            _projectServices = projectServices;
+            _projectLockService = projectLockService;
+        }
+
+        /// <summary>
+        /// Called on project load.
+        /// </summary>
+        [ConfiguredProjectAutoLoad]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+        internal void Load()
+        {
+            this.EnsureInitialized();
+        }
+        
+        /// <summary>
+        /// Initialize the watcher.
+        /// </summary>
+        protected override void Initialize()
+        {
+            _treeWatcher = _fileSystemTreeProvider.Tree.LinkTo(new ActionBlock<IProjectVersionedValue<IProjectTreeSnapshot>>(new Action<IProjectVersionedValue<IProjectTreeSnapshot>>(this.ProjectTree_ChangedAsync)));
+        }
+
+        /// <summary>
+        /// Called on changes to the project tree.
+        /// </summary>
+        internal void ProjectTree_ChangedAsync(IProjectVersionedValue<IProjectTreeSnapshot> treeSnapshot)
+        {
+            var newTree = treeSnapshot.Value.Tree;
+            if (newTree == null)
+            {
+                return;
+            }
+
+            // If there' no project.json in the project, there's nothing to watch.
+            IProjectTree projectJsonNode = FindProjectJsonNode(newTree);
+            if (projectJsonNode == null)
+            {
+                // project.json may have been deleted.
+                UnregisterFileWatcherIfAny();
+                return;
+            }
+
+            var projectDirectory = Path.GetDirectoryName(_projectServices.Project.FullPath);
+            var projectLockJsonFilePath = Path.ChangeExtension(PathHelper.Combine(projectDirectory, projectJsonNode.Caption), ".lock.json");
+
+            // project.json may have been renamed to {projectName}.project.json. In that case change the file watcher.
+            if (!PathHelper.IsSamePath(projectLockJsonFilePath, _fileBeingWatched))
+            {
+                UnregisterFileWatcherIfAny();
+                RegisterFileWatcher(projectLockJsonFilePath);
+                _fileBeingWatched = projectLockJsonFilePath;
```

Comment:
Is this event guaranteed to never overlap so that you have multiple threads in this method at once?